# House Prices: Advanced Regression Techniques
https://www.kaggle.com/c/house-prices-advanced-regression-techniques/overview

## Load data

In [1]:
import pandas as pd

df = pd.read_csv('data/train.csv')

df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities    ...     PoolArea PoolQC Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  

[5 rows x 81 columns]

## Feature Engineering

### Check not number columns 

In [21]:
not_number_columns = df.columns[df.dtypes == object]
not_number_columns

Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature',
       'SaleType', 'SaleCondition'],
      dtype='object')

In [23]:
print("Number of not number columns:", not_number_columns.size)

Number of not number columns: 43


### Map them to N-1 binary columns

Calculate how many unique values has each not number column

In [28]:
number_of_new_columns = 0
for col_name in not_number_columns:
    nunique = df[col_name].nunique()
    number_of_new_columns += nunique-1
    print(col_name, " - ", nunique)

print("\nNew columns:", number_of_new_columns)

MSZoning  -  5
Street  -  2
Alley  -  2
LotShape  -  4
LandContour  -  4
Utilities  -  2
LotConfig  -  5
LandSlope  -  3
Neighborhood  -  25
Condition1  -  9
Condition2  -  8
BldgType  -  5
HouseStyle  -  8
RoofStyle  -  6
RoofMatl  -  8
Exterior1st  -  15
Exterior2nd  -  16
MasVnrType  -  4
ExterQual  -  4
ExterCond  -  5
Foundation  -  6
BsmtQual  -  4
BsmtCond  -  4
BsmtExposure  -  4
BsmtFinType1  -  6
BsmtFinType2  -  6
Heating  -  6
HeatingQC  -  5
CentralAir  -  2
Electrical  -  5
KitchenQual  -  4
Functional  -  7
FireplaceQu  -  5
GarageType  -  6
GarageFinish  -  3
GarageQual  -  5
GarageCond  -  5
PavedDrive  -  3
PoolQC  -  3
Fence  -  4
MiscFeature  -  4
SaleType  -  9
SaleCondition  -  6

New columns: 209


Not looks good. If we use map them to N-1 binary columns, we will have a new 209 columns and 289 features.

*TODO: Add map to N-1 binary columns*

In [2]:
print("Number of rows:", df.shape[0])
print("Number of features:", df.shape[1])

Number of rows: 1460
Number of features: 81


The data set contains 80 source features and only 1460 examples. The ratio features to examples aren't so good, so the plan is the limit of the features, maybe some features can be omitted. Firstly start with a genetic algorithm for select features and nextly use SVM algorithm.

## Split data 
Test data frame - 20%
Train data frame - 80%

In [3]:
from sklearn.model_selection import train_test_split

train_df, test_df = train_test_split(df, test_size=0.2)

In [5]:
X_train = train_df.drop(['SalePrice'], axis=1)
y_train = train_df['SalePrice']

X_test = test_df.drop(['SalePrice'], axis=1)
y_test = test_df['SalePrice']

print("Number of training samples: ", X_train.shape[0])
print("Number of test samples: ", X_test.shape[0])

Number of training samples:  1168
Number of test samples:  292


## Support Vector Machine Algorithm
Before start working with a genetic algorithm, we need to try to find the best hyperparameters for SVM and all features

TBA